### PR TITLE
Add thumbnailer

### LINKED ISSUE #143: Create thumbnails on demand
Generate a thumbnail for a given pipeline when requested.

reference: [thumbnail.rs](https://gitlab.freedesktop.org/gstreamer/gstreamer-rs/-/blob/main/examples/src/bin/thumbnail.rs)

### LINKED ISSUE #161: Can't create UDP and RTSP streams at the same time
If we use endpoints

`udp://0.0.0.0:5600, udp://0.0.0.0:5601` creates one `multiudpsink`, but when we add an RTSP, like `udp://0.0.0.0:5600, udp://0.0.0.0:5601, rtsp://0.0.0.0:8554/test`, then it fails.

Sinks are added to the pipeline:
image

Relevant Log:
[gst_debug=4.txt](https://github.com/bluerobotics/mavlink-camera-manager/files/10261814/gst_debug.4.txt)
[dots.zip](https://github.com/bluerobotics/mavlink-camera-manager/files/10261833/dots.zip)

### PR DESCRIPTION
**Pull request by joaoantoniocardoso**
Closes #143
Fixes #161 (the underlying problem, although RTSP + UDP is kept disabled)

### CODE REVIEW

File: src/stream/mod.rs [hunk 1]
```diff
@@ -87,6 +87,15 @@ impl Stream {
             }
         }
 
+        if let Err(reason) =
```
Comment (patrickelectric) [inline src/stream/mod.rs:90]: For the future reference, we should think how stable is this and add a command line argument to enable and disable this functionality.

File: src/server/pages.rs [hunk 2]
```diff
@@ -304,3 +311,31 @@ pub fn sdp(sdp_file_request: web::Query<SdpFileRequest>) -> HttpResponse {
             )),
     }
 }
+
+#[api_v2_operation]
+/// Provides a sdp description file that contains information for a specific stream, based on: [RFC 8866](https://www.rfc-editor.org/rfc/rfc8866.html)
```
Comment (patrickelectric) [inline src/server/pages.rs:316]: The description is wrong.

File: src/server/pages.rs [hunk 3]
```diff
@@ -68,6 +68,13 @@ pub struct SdpFileRequest {
     source: String,
 }
 
+#[derive(Apiv2Schema, Debug, Deserialize)]
+pub struct ThumbnailFileRequest {
+    source: String,
+    quality: Option<u8>,
```
Comment (patrickelectric) [inline src/server/pages.rs:74]: What quality is not clear, can you add a comment ?

File: src/stream/sink/image_sink.rs [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+                .collect::<Vec<&gst::Element>>(),
+        );
+        elements.push(self.appsink.upcast_ref());
+        let elements = &elements;
+        if let Err(error) = pipeline.add_many(elements) {
+            return Err(anyhow!(
+                "Failed adding ImageSink {sink_id} elements to Pipeline {pipeline_id}. Reason: {error:?}"
+            ));
+        }
+
```
[the remark this answers is not in the document]
Comment (joaoantoniocardoso) [reply src/stream/sink/image_sink.rs:57]: We actually don't need any of these dotfiles here, were used during development, but don't help much for debugging, if needed, I can add them later for all sinks.

File: src/stream/sink/image_sink.rs [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+                            channel_stride: 1, // 1 byte from component to component
+                            width: frame.width(),
+                            width_stride: 4, // 4 byte from pixel to pixel
+                            height: frame.height(),
+                            height_stride: frame.plane_stride()[0] as usize, // stride from line to line
+                        },
+                        color_hint: Some(image::ColorType::Rgb8),
+                    });
+
+                // std::thread::sleep(std::time::Duration::from_secs(1)); // This defines the interval between shots
```
[the remark this answers is not in the document]
Comment (joaoantoniocardoso) [reply src/stream/sink/image_sink.rs:266]: oops, this shouldn't be disabled :eyes:

File: src/stream/mod.rs (hunk 1 above)
Comment (joaoantoniocardoso) [reply src/stream/mod.rs:90]: true

File: src/stream/sink/image_sink.rs [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+
+                // Create a FlatSamples around the borrowed video frame data from GStreamer with
+                // the correct stride as provided by GStreamer.
+                flat_samples_cloned
+                    .lock()
+                    .unwrap()
+                    .replace(image::FlatSamples::<Vec<u8>> {
+                        samples: frame.plane_data(0).unwrap().to_vec(),
+                        layout: image::flat::SampleLayout {
+                            channels: 3,       // RGB
```
[the remark this answers is not in the document]
Comment (joaoantoniocardoso) [reply src/stream/sink/image_sink.rs]: yeah, its the cargo format default to align consecutive comments, I changed to the previous line

File: src/server/pages.rs (hunk 3 above)
Comment (patrickelectric) [reply src/server/pages.rs:75]: You should use a validator here for the range.
https://github.com/Keats/validator

Comment (joaoantoniocardoso) [reply src/server/pages.rs:78]: Done, after a fight with versioning

File: Cargo.toml [hunk 7]
```diff
@@ -27,18 +27,20 @@ regex = "1.6.0"
...
-actix-service = "2.0.2"
-serde = { version = "1.0.140", features = ["derive"] }
-serde_json = "1.0.82"
+actix-files = "0.6"
+actix-web = "4"
+actix-web-validator = "5"
+actix-service = "2"
+serde = { version = "1", features = ["derive"] }
+serde_json = "1"
+validator = { version = "0.16", features = ["derive"] }
```
Comment (joaoantoniocardoso) [inline Cargo.toml:36]: Should we stick to a minor version for these libs?

Comment (patrickelectric) [reply Cargo.toml]: That's not necessary, minor versions does not break apis, just bug fixes and improvements

**Comment by joaoantoniocardoso** [comment]
For reference, on the raspbery pi 4, the ImageSink (thumbnailer) costs around +2% of CPU for an H264 1080p streams and takes less than 50ms to respond a 240p thumbnail request.

**Review by patrickelectric — APPROVED** [review]

**Pull request merged**